test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java

```diff
@@ -0,0 +1,253 @@
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package compiler.c2.loopopts;
+
+import compiler.lib.ir_framework.*;
+import jdk.test.lib.Asserts;
+
+/*
+ * @test
+ * @bug 8323220
+ * @summary Test loop invariant code motion for cmp nodes through reassociation
+ * @library /test/lib /
+ * @run driver compiler.c2.loopopts.InvariantCodeMotionReassociateCmp
+ */
+public class InvariantCodeMotionReassociateCmp {
+
+    public static void main(String[] args) {
+        TestFramework.run();
+    }
+
+    @DontInline
+    private void blackhole() {}
+
+    @Test
+    @Arguments({Argument.RANDOM_EACH, Argument.RANDOM_EACH})
+    @IR(counts = {IRNode.SUB_I, "1"})
+    public void equalsAddInt(int inv1, int inv2) {
+        for (int i = 0; i < 500; ++i) {
+            // Reassociate to `inv2 - inv1 == i`
+            if (inv1 + i == inv2) {
+                blackhole();
+            }
+        }
+    }
+
+    @Test
+    @Arguments({Argument.RANDOM_EACH, Argument.RANDOM_EACH})
+    @IR(counts = {IRNode.SUB_L, "1"})
+    public void equalsAddLong(long inv1, long inv2) {
+        for (int i = 0; i < 500; ++i) {
+            // Reassociate to `inv2 - inv1 == i`
+            if (inv1 + i == inv2) {
+                blackhole();
+            }
+        }
+    }
+
+    @Test
+    @Arguments({Argument.NUMBER_42, Argument.NUMBER_42})
+    @IR(counts = {IRNode.SUB_I, "1"})
+    public void equalsInvariantSubVariantInt(int inv1, int inv2) {
+        for (int i = 0; i < 500; ++i) {
+            // Reassociate to `inv1 - inv2 == i`
+            if (inv1 - i == inv2) {
+                blackhole();
+            }
+        }
+    }
+
+    @Test
+    @Arguments({Argument.NUMBER_42, Argument.NUMBER_42})
+    @IR(counts = {IRNode.SUB_L, "1"})
+    public void equalsInvariantSubVariantLong(long inv1, long inv2) {
+        for (int i = 0; i < 500; ++i) {
+            // Reassociate to `inv1 - inv2 == i`
+            if (inv1 - i == inv2) {
+                blackhole();
+            }
+        }
+    }
+
+    @Test
+    @Arguments({Argument.NUMBER_42, Argument.NUMBER_42})
+    @IR(counts = {IRNode.ADD_I, "2"})
+    public void equalsVariantSubInvariantInt(int inv1, int inv2) {
+        for (int i = 0; i < 500; ++i) {
+            // Reassociate to `inv1 + inv2 == i`
+            if (i - inv1 == inv2) {
+                blackhole();
+            }
+        }
+    }
+
+    @Test
+    @Arguments({Argument.NUMBER_42, Argument.NUMBER_42})
+    @IR(counts = {IRNode.ADD_L, "1"})
+    public void equalsVariantSubInvariantLong(long inv1, long inv2) {
+        for (int i = 0; i < 500; ++i) {
+            // Reassociate to `inv1 + inv2 == i`
+            if (i - inv1 == inv2) {
+                blackhole();
+            }
+        }
+    }
```

Comment:
I suggest you add a test that first iterates some number of iterations, and then returns inside the if-statement with return value i. Then you check if that value is as expected.
Because with all the tests that you have here, I'm not sure if really all branches are always taken, or if they may lead to uncommon traps, and hence leave some things uncovered by testing.

Reply:
@eme64 can you take a look at the updates?

Reply:
I don't think you have added what I requested. Do you have any test that does the reassociation, and where you verify that the computed values are correct?

Reply:
OK thanks I think I misread. I now have some correctness tests for tests with and without reassociation. 

We can't add correctness tests for tests with random inputs because we do not know how many iterations there will be. There are two tests with reassociation `equalsVariantSubInvarint(Int|Long)` that have a drastically different graph with the correctness change that messes up the IR matching, so I opted out of changes those as well.

Reply:
In fact, that would probably be smart. Otherwise you only have result verification for constant values. And who knows, there the code might constant-fold in a way that we don't expect.

Reply:
I am with you that I am not 100% happy with the tests. I do not mind waiting for that PR to get integrated.

If I understand correctly, that PR is more for convenience of setting up arguments. How will that help testing correctness for random inputs?

Reply:
I added correctness tests for two of the tests with random inputs. For the rest, it messes up the IR matching. This goes back to the fragility of matching add/sub nodes.